Add new entities of RW content types to default stage

jhoward1994 opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

Decorates the entity service to add new entities of RW content types to the default workflow stage.

### Why is it needed?

Entity creation doesn't trigger server restart so EE bootstrap will not pick up these new entities

### How to test it?

WIP

### Related issue(s)/PR(s)

CONTENT-1180

jhoward1994 commented:
@Marc-Roig @nathan-pichon This PR adds new entities of RW content types to the default workflow stage without needing the bootstrap phase to run.

I'm unsure about something: the entity service is now decorated from both i18n and EE. How will this behave when the two interact? It seems to be working fine for me but I don't have a good understanding of how the decoration works

Marc-Roig commented on packages/core/admin/ee/server/services/review-workflows/entity-service-decorator.js:
```diff
@@ -0,0 +1,30 @@
+ * @param {object} service - entity service
+ */
+const decorator = (service) => ({
+  async create(uid, opts = {}) {
+    const model = strapi.getModel(uid);
+    const hasRW = hasRWEnabled(model);
+
+    const entity = await service.create.call(this, uid, opts);
+    if (!hasRW) {
+      return entity;
+    }
+
+    // Assign this entity to the default workflow stage
+    const { assignEntityDefaultStage } = getService('review-workflows');
+    await assignEntityDefaultStage(uid, entity.id);
+
+    return entity;
+  },
```
If I understand correctly the entity is created and then assigned the default workflow if it's not assigned no?

I see a couple of issues/inconveniences with this approach:
- The triggered webhooks and lifecycles events in the `service.create.call` will not have the stage information.
- We are making additional database queries if RW is enabled.

I can see this minor problems fixed if we do something similar as i18n:
```js
  // packages/plugins/i18n/server/services/entity-service-decorator.js
  async create(uid, opts = {}) {
    const model = strapi.getModel(uid);

    const { syncLocalizations, syncNonLocalizedAttributes } = getService('localizations');
    const { isLocalizedContentType } = getService('content-types');

    if (!isLocalizedContentType(model)) {
      return service.create.call(this, uid, opts);
    }

    const { data } = opts;
    await assignValidLocale(data); // <==== We assign the locale directly in the parameters if it is not set

    const entry = await service.create.call(this, uid, opts);
	...
    return entry;
  },
  ```

Do you think this could be possible here?

Marc-Roig commented:
@jhoward1994 
> I'm unsure about something: the entity service is now decorated from both i18n and EE. How will this behave when the two interact? It seems to be working fine for me but I don't have a good understanding of how the decoration works

It should not be a problem actually! The decorator wraps the method so decorating it multiple times should just wrap it multiple times. 

[The implementation of the decorator pattern](https://github.com/strapi/strapi/blob/main/packages/core/strapi/lib/services/entity-service/index.js#L303):
```
  const service = {
    implementation,
    decorate(decorator) {
      if (typeof decorator !== 'function') {
        throw new Error(`Decorator must be a function, received ${typeof decorator}`);
      }

      this.implementation = { ...this.implementation, ...decorator(this.implementation) };
      return this;
    },
  };
```

Marc-Roig commented:
🚀 Well done, I have added a couple of comments

nathan-pichon commented on packages/core/admin/ee/server/services/review-workflows/review-workflows.js:
```diff
@@ -157,18 +157,48 @@ function enableReviewWorkflow({ strapi }) {
+ */
+async function assignEntityDefaultStage(uid, entityID) {
+  const defaultWorkflow = await getDefaultWorkflow({ strapi });
+  if (!defaultWorkflow) {
+    return;
+  }
+  const firstStage = defaultWorkflow.stages[0];
```
Should not be possible

jhoward1994 commented:
Thanks both for the review I've addressed the feedback in [`9a43416` (#16122)](https://github.com/strapi/strapi/pull/16122/commits/9a4341639c0fe32e2434bc4676bbd0a07b38de46)

ping @Marc-Roig @nathan-pichon

Marc-Roig commented on packages/core/admin/ee/server/services/review-workflows/entity-service-decorator.js:
```diff
@@ -0,0 +1,30 @@
+      return service.create.call(this, uid, opts);
+    }
+
+    const defaultWorkFlow = await getDefaultWorkflow({ strapi });
+
+    return service.create.call(this, uid, {
+      ...opts,
+      // Assign this entity to the default workflow stage
+      data: { ...opts.data, strapi_reviewWorkflows_stage: defaultWorkFlow.stages[0].id },
+    });
```
I would check if `opts.data` already has a `stage` assigned. 

If a user wants to set a specific stage, this would completely override the value.
```js
const data = { ...opts.data }

if (!has('strapi_reviewWorkflows_stage', data)) {
  data.strapi_reviewWorkflows_stage = await getDefaultWorkflow({ strapi });
 }
```

Marc-Roig replied:
or maybe using isNil instead of has🤔 so we don't allow the value to be nullish

nathan-pichon replied:
You could do another func 👍 
```javascript
async function getDataWithStage({ strapi }, data) {
  if (!has(ENTITY_ATTRIBUTE_NAME, data)) {
    const defaultWorkflow = await getDefaultWorkflow({ strapi });
    return { ...data, [ENTITY_ATTRIBUTE_NAME]: defaultWorkflow.stages[0].id };
  }
  return data;
}
```

jhoward1994 replied:
Thanks 🙏🏻 [`0c1f8a3` (#16122)](https://github.com/strapi/strapi/pull/16122/commits/0c1f8a3115e4bc9e83ea39a6a1f822734c940bc6)

Marc-Roig commented on packages/core/admin/ee/server/tests/review-workflows.test.api.js:
```diff
@@ -373,4 +373,38 @@ describeOnCondition(edition === 'EE')('Review workflows', () => {
+
+      expect(response.body.data.schema.reviewWorkflows).toBeTruthy();
+
+      response = await getRWMorphTableResults(strapi.db.getConnection());
+
+      expect(response.length).toEqual(6);
+      for (let i = 0; i < response.length; i += 1) {
+        const entry = response[i];
+        expect(entry.related_id).toEqual(i + 1);
+        expect(entry.order).toEqual(1);
+      }
```
last comment I promise 🙈 

Why not looking at the response of the admin api? it shoudl contain the `strapi_reviewWorkflow_stage` attribute. 

That way this test doesnt have to know that attribute is using a morph table.

Marc-Roig reviewed (approved):
LGTM
